### PR TITLE
Fix: SAML group_mapping update issue, and adding issuer_url support

### PR DESCRIPTION
**Pull request by jschoombee**
### Background 

As mentioned in [this](https://github.com/AxtonGrams/terraform-provider-wiz/issues/39) issue. The Wiz mutation for  `UpdateSAMLIdentityProvider` requires all fields, once set, the key needs to be present in subsequent requests, i.e. `update`. As described in the issue, this is experienced when trying to make changes (`patch`) existing `group_mappings` With the current stable release, it is also not possible for example, to update or patch values like `login_url`.  

The ability to set the `issuer_url` was also not yet implemented.

### Proposed Changes
- Allow the setting of the `issuer_url` so that it does not automatically get set to the `login` url but can also support specific non URL values such as `URIs`/`ARNs`
- Updated example HCL file for `wiz_saml_idp`
- Updated `wiz_saml_idp` schema attribute dependencies
- Marked required fields

### CODE REVIEW

File: internal/provider/resource_saml_idp.go [hunk 1]
```diff
@@ -23,6 +23,11 @@ func resourceWizSAMLIdP() *schema.Resource {
 				Description: "Internal identifier for the Saml Provider",
 				Computed:    true,
 			},
+			"issuer_url": {
+				Type:        schema.TypeString,
+				Description: "IdP issuer URL or entity ID. The value needs to match the IdP Issuer value, which *should* generally be a URL but can also be a URI or simple string. If unsure, set to same as login_url.",
+				Required:    true,
```
Comment (gramsa49) [inline internal/provider/resource_saml_idp.go:29]: This field is optional, per the GraphQL schema.

```
type SAMLIdentityProvider {
  allowManualRoleOverride: Boolean!
  certificate: String!
  domains: [String!]
  groupMapping: [SAMLGroupMapping!]
  id: ID!
  issuerURL: String
  loginURL: String!
  logoutURL: String!
  mergeGroupsMappingByRole: Boolean!
  name: String!
  useProviderManagedRoles: Boolean!
}
```

File: internal/provider/resource_saml_idp.go [hunk 2]
```diff
@@ -40,18 +45,18 @@ func resourceWizSAMLIdP() *schema.Resource {
 			},
 			"use_provider_managed_roles": {
 				Type:        schema.TypeBool,
-				Description: "Use provider managed roles?",
-				Optional:    true,
-				Default:     false,
+				Description: "Use provider (IdP) managed roles? This attribute needs to be set to accomodate for update lifecycle operations.",
+				Required:    true,
```
Comment (gramsa49) [inline internal/provider/resource_saml_idp.go:49]: Though the GraphQL schema labels this as required, this should remain optional with a default of false.  Default values cannot be defined for required attributes.

File: internal/provider/resource_saml_idp.go [hunk 3]
```diff
@@ -40,18 +45,18 @@ func resourceWizSAMLIdP() *schema.Resource {
...
 			},
 			"allow_manual_role_override": {
 				Type:        schema.TypeBool,
-				Description: "Allow manual override for role assignment? Must be set `true` if `use_provided_roles` is false.",
-				Optional:    true,
-				Default:     true,
-				RequiredWith: []string{
-					"use_provider_managed_roles",
-				},
```
Comment (gramsa49) [inline internal/provider/resource_saml_idp.go:54]: Though the GraphQL schema labels this as required, this should remain optional with a default of true. Default values cannot be defined for required attributes. Changing the default behavior at this stage may have unintended consequences for users of the provider.

File: internal/provider/resource_saml_idp.go (hunk 1 above)
Comment (jschoombee) [reply internal/provider/resource_saml_idp.go:29]: Indeed it should be optional. This speaks to the bug that Wiz introduced when moving to Cognito which is, according to support, limited to this specific mutation in that all attributes once passed, need to be passed for the lifecycle. 

Assuming the value gets set at creation but unset at some stage, the API would return an error which is what was tested.

File: internal/vendor/wiz.go [hunk 4]
```diff
@@ -177,15 +177,15 @@ type UpdateSAMLIdentityProviderInput struct {
...
-	MergeGroupsMappingByRole *bool                         `json:"mergeGroupsMappingByRole,omitempty"`
+	LoginURL                 string                        `json:"loginURL"`
+	LogoutURL                string                        `json:"logoutURL"`
+	IssuerURL                string                        `json:"issuerURL"`
+	UseProviderManagedRoles  *bool                         `json:"useProviderManagedRoles"`
+	AllowManualRoleOverride  *bool                         `json:"allowManualRoleOverride"`
+	Certificate              string                        `json:"certificate"`
+	Domains                  []string                      `json:"domains"`
+	GroupMapping             []SAMLGroupMappingUpdateInput `json:"groupMapping"`
+	MergeGroupsMappingByRole *bool                         `json:"mergeGroupsMappingByRole"`
```
Comment (gramsa49) [inline internal/vendor/wiz.go:188]: The omitempty tag is based on the GraphQL schema.  The golang struct should mirror the GraphQL schema where possible.  For this struct, all elements are optional.

```
input UpdateSAMLIdentityProviderPatch {
  name: String
  issuerURL: String
  loginURL: String
  logoutURL: String
  useProviderManagedRoles: Boolean
  allowManualRoleOverride: Boolean
  certificate: String
  domains: [String!]
  groupMapping: [SAMLGroupMappingUpdateInput!]
  mergeGroupsMappingByRole: Boolean
}
```

File: internal/provider/resource_saml_idp.go (hunk 2 above)
Comment (jschoombee) [reply internal/provider/resource_saml_idp.go:49]: The same comment as above, I was hoping there's a better approach here than to make all values default because the schema is not respected for patch ops once change gets introduced

File: internal/provider/resource_saml_idp.go [hunk 5]
```diff
@@ -229,6 +230,7 @@ func resourceWizSAMLIdPRead(ctx context.Context, d *schema.ResourceData, m inter
 	        allowManualRoleOverride
 	        certificate
 	        domains
+			issuerURL
```
Comment (gramsa49) [inline internal/provider/resource_saml_idp.go:233]: Indentation on the graphql query is a tab for code indentatoin, followed by spaces for query indentation.  To keep things consistent, this formatting should be maintained.

File: internal/vendor/wiz.go [hunk 6]
```diff
@@ -203,15 +203,15 @@ type SAMLGroupMappingUpdateInput struct {
 // CreateSAMLIdentityProviderInput struct -- updates
 type CreateSAMLIdentityProviderInput struct {
```
Comment (gramsa49) [inline internal/vendor/wiz.go:204]: The golang struct should mirror the GraphQL schema when deciding whether to omit empty attributes.  E.g., allowManualRoleOverride is optional, so the golang struct should have the omitempty tag.

```
input CreateSAMLIdentityProviderInput {
  name: String!
  issuerURL: String
  loginURL: String!
  logoutURL: String
  useProviderManagedRoles: Boolean!
  allowManualRoleOverride: Boolean
  certificate: String!
  domains: [String!]
  groupMapping: [SAMLGroupMappingCreateInput!]
  mergeGroupsMappingByRole: Boolean
}
```

File: internal/vendor/wiz.go [hunk 7]
```diff
@@ -230,9 +230,9 @@ type SAMLGroupMappingCreateInput struct {
 type SAMLIdentityProvider struct {
 	AllowManualRoleOverride  *bool               `json:"allowManualRoleOverride"`
 	Certificate              string              `json:"certificate"`
-	Domains                  []string            `json:"domains"`
-	EntityID                 string              `json:"entityID,omitempty"`
-	GroupMapping             []*SAMLGroupMapping `json:"groupMapping,omitempty"`
+	Domains                  []*string           `json:"domains"`
+	IssuerURL                string              `json:"issuerURL"`
+	GroupMapping             []*SAMLGroupMapping `json:"groupMapping"`
```
Comment (gramsa49) [inline internal/vendor/wiz.go:235]: The golang struct should mirror the GraphQL schema (omitempty).  issuerURL and groupMapping allow empty values.

```
type SAMLIdentityProvider {
  allowManualRoleOverride: Boolean!
  certificate: String!
  domains: [String!]
  groupMapping: [SAMLGroupMapping!]
  id: ID!
  issuerURL: String
  loginURL: String!
  logoutURL: String!
  mergeGroupsMappingByRole: Boolean!
  name: String!
  useProviderManagedRoles: Boolean!
}
```

File: internal/vendor/wiz.go [hunk 8]
```diff
@@ -230,9 +230,9 @@ type SAMLGroupMappingCreateInput struct {
 type SAMLIdentityProvider struct {
 	AllowManualRoleOverride  *bool               `json:"allowManualRoleOverride"`
 	Certificate              string              `json:"certificate"`
-	Domains                  []string            `json:"domains"`
-	EntityID                 string              `json:"entityID,omitempty"`
-	GroupMapping             []*SAMLGroupMapping `json:"groupMapping,omitempty"`
+	Domains                  []*string           `json:"domains"`
```
[the remark this answers is not in the document]
Comment (jschoombee) [reply internal/vendor/wiz.go:233]: was something still on the `to-do, there is a bug in currently release that the slice for domains is sent as `null`, so speculative plans would constantly see it is needed change

File: internal/provider/resource_saml_idp.go [hunk 9]
```diff
@@ -61,7 +66,7 @@ func resourceWizSAMLIdP() *schema.Resource {
 			"domains": {
 				Type:        schema.TypeList,
 				Required:    true,
-				Description: "A list of domains the IdP handles.",
+				Description: "A list of domains the IdP handles. This block needs to be set to accomodate for update lifecycle operations. If not used then specify an empty list (`[]`)",
```
Comment (gramsa49) [inline internal/provider/resource_saml_idp.go:69]: It looks like domain can be made optional.  In prior runs, I recall the field being required, but now it appears to be optional.

File: internal/provider/resource_saml_idp.go [hunk 10]
```diff
@@ -40,18 +45,18 @@ func resourceWizSAMLIdP() *schema.Resource {
...
-				RequiredWith: []string{
-					"use_provider_managed_roles",
-				},
+				Required:    true,
+				Description: "Allow manual override for role assignment? This attribute must be set to `true` if `use_provider_managed_roles` attribute is set to `false`. This field needs to be set to accomodate for update lifecycle operations.",
+			},
+			"merge_groups_mapping_by_role": {
+				Type:        schema.TypeBool,
+				Description: "Manage group mapping by role? This attribute needs to be set to accomodate for update lifecycle operations.",
+				Required:    true,
```
Comment (gramsa49) [inline internal/provider/resource_saml_idp.go:59]: Why was this changed from optional to required?  The GraphQL schema indicates that this is optional.

```
input CreateSAMLIdentityProviderInput {
  name: String!
  issuerURL: String
  loginURL: String!
  logoutURL: String
  useProviderManagedRoles: Boolean!
  allowManualRoleOverride: Boolean
  certificate: String!
  domains: [String!]
  groupMapping: [SAMLGroupMappingCreateInput!]
  mergeGroupsMappingByRole: Boolean
}
```

File: internal/provider/resource_saml_idp.go (hunk 9 above)
Comment (jschoombee) [reply internal/provider/resource_saml_idp.go:69]: See https://github.com/AxtonGrams/terraform-provider-wiz/pull/42#discussion_r1036255152

File: internal/provider/resource_saml_idp.go (hunk 10 above)
Comment (jschoombee) [reply internal/provider/resource_saml_idp.go:59]: see https://github.com/AxtonGrams/terraform-provider-wiz/pull/42#discussion_r1036255152

File: docs/resources/saml_idp.md [hunk 11]
```diff
@@ -17,8 +17,10 @@ resource "wiz_saml_idp" "test" {
   name                         = "Ping"
   login_url                    = "https://ping.example.com/idp/SSO.saml2"
   logout_url                   = "https://ping.example.com/idp/SLO.saml2"
+  issuer_url                   = "urn:mace:incommon:example.edu"
```
Comment (gramsa49) [inline docs/resources/saml_idp.md:20]: The docs are auto-generated from the examples and provider resource schemas.  To change the examples, update the relavent example in ./examples/.

File: internal/provider/resource_saml_idp.go (hunk 1 above)
Comment (gramsa49) [reply internal/provider/resource_saml_idp.go:29]: Thanks for the clarification.  We can work around this with the golang struct.  The Resource schema can still make this optional.  Make sense?

File: internal/provider/resource_saml_idp.go (hunk 2 above)
Comment (gramsa49) [reply internal/provider/resource_saml_idp.go:49]: With this, we can achieve the desired effect with modification to the golang struct, such that empty values are not omitted.  The changes to the struct can be limited in scope to UpdateSAMLIdentityProviderPatch.

I ran into similar issues with other mutations.

File: internal/provider/resource_saml_idp.go (hunk 1 above)
Comment (gramsa49) [reply internal/provider/resource_saml_idp.go:29]: I created a saml idp without the issuerurl defined, but when reading the resource, it appears the issuerurl was set to the loginurl:

Create variables:

```
{
  "input": {
    "name": "test-bool",
    "loginURL": "https://example.com/idp/SSO.saml2",
    "logoutURL": "https://example.com/idp/SLO.saml2",
    "useProviderManagedRoles": true,
    "allowManualRoleOverride": true,
    "certificate": "-----BEGIN ..."
    "domains": [],
    "mergeGroupsMappingByRole": false
  }
}
```

Read response:

```
{
  "data": {
    "samlIdentityProvider": {
      "id": "test-bool",
      "name": "test-bool",
      "issuerURL": "https://example.com/idp/SSO.saml2",
      "loginURL": "https://example.com/idp/SSO.saml2",
      "logoutURL": "https://example.com/idp/SLO.saml2",
      "useProviderManagedRoles": true,
      "allowManualRoleOverride": true,
      "domains": null,
      "mergeGroupsMappingByRole": false,
      "groupMapping": []
    }
  }
}

Comment (jschoombee) [reply internal/provider/resource_saml_idp.go:29]: Indeed thats Wiz's default behaviour to set to loginURL if undefined. I tested when creating it in TF as undefined (when it was optional) the create would work, but the update would fail as it wasn't part of the original fields set at creation hence why this one also made required to work around.

Comment (jschoombee) [reply internal/provider/resource_saml_idp.go:29]: Will test this then if the read response updates state why I found above behaviour..\

File: internal/provider/resource_saml_idp.go (hunk 2 above)
Comment (gramsa49) [reply internal/provider/resource_saml_idp.go:49]: Ok, so I tracked down where I ran into a similar issue (different provider).  I used the [GetChange](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.1/helper/schema#ResourceData.GetChange) function to get the old and new value, then handled the updates.

```
func accountUpdateString(key, cakey string, d *schema.ResourceData) interface{} {
	log.Printf("accountUpdate called for resource %s", d.Id())
	old, new := d.GetChange(key)
	_ = old
	if old == "" {
		update := &vendor.AccountUpdateString{
			Op:    "add",
			Path:  cakey,
			Value: new.(string),
		}
		return update
	} else if new == "" {
		update := &vendor.AccountUpdateRemove{
			Op:   "remove",
			Path: cakey,
		}
		return update
	} else {
		update := &vendor.AccountUpdateString{
			Op:    "replace",
			Path:  cakey,
			Value: new.(string),
		}
		return update
	}
}
```

We should do something similar to handle cases during update where a field was nullified or set to an empty value.  If we couple that with a deviation in struct UpdateSAMLIdentityProviderPatch, we can pass empty strings in the update variables, which should have the desired effect.
